M2/Macaulay2/packages/GKMManifolds.m2

```diff
@@ -0,0 +1,1468 @@
+
+--input: a list L = {l_1, ... , l_d} of list of integers, or
+--a matrix A whose columns are l_i's, representing exponents of a monomial map
+--output: a quotient ring R/I, where R has variables x_1, ... x_d of degrees l_1, ... , l_d and
+--I is the toric ideal defined by the monomial map
+affineToricRing = method();
+affineToricRing(List) := QuotientRing => L -> (
+    d := #L;
+    x := symbol x;
+    R := QQ[apply(#L, i -> x_i), Degrees => L];
+    I := if #L == 0 then ideal(0_R) else toricGroebner(transpose matrix L, R);
+    R/I
+    )
+
+affineToricRing(Matrix) := QuotientRing => A -> affineToricRing(entries transpose A)
+
+--internal auxiliary function:
+--default Macaulay2 can't do fraction fields for Laurent rings
+--but by shifting so that we exit Laurent rings and then going back, we can.
+--Here given two Laurent polynomials f,g and a ring S with Inverses=>false, outputs the ratio
+--f/g of f and g by multiplying both of them with a big enough monomial to make them polynomials
+--the output is the ratio in the new ring S,
+--and a function goBack which puts the element in the ring of f if the denominator is a monomial
+toFraction = method();
+toFraction(RingElement,RingElement,Ring) := (f,g,S) -> (
+    R := ring f;
+    if not R === ring g then << " the two polynomials live in different rings " << return error;
+    if not (#gens R) == (#gens S) then << " the temporary ring is no good " << return error;
+    Exps := -((transpose (exponents f | exponents g))/min);
+    mult := product(#Exps, i -> R_i^(Exps_i));
+    numer := sub(mult*f, apply(#(gens R), i -> R_i=>S_i));
+    denom := sub(mult*g, apply(#(gens R), i -> R_i=>S_i));
+    rat := numer / denom;
+    goBack := ratFct -> (
+	N := numerator ratFct;
+	D := denominator ratFct;
+	if not #(terms D) == 1 then << "denominator not a monomial" << return error;
+	sub(N, apply(#(gens S), i -> S_i=>R_i))*(sub(D, apply(#(gens S), i -> S_i=>R_i)))^(-1)
+	);
+    {rat, goBack}
+)
+
+--auxiliary function:
+--makes a polynomial ring with n variables, representing T-equivariant cohomology 
+--ring of a point where T =  = (kk^*)^n.
+makeHTpt = method()
+makeHTpt(ZZ) := Ring => n -> (
+    t := symbol t;
+    QQ[t_0..t_(n-1)]
+    )
+
+
+-----------------------------------------------------------------------------------------------
+--------------------------------------< moment graphs >----------------------------------------
+-----------------------------------------------------------------------------------------------
+
+--a MomentGraph G is a MutableHashTable with the minimal data of:
+--G.vertices : a list representing vertices of the graph G
+--G.edges : a hash table whose keys are pairs {p,q} of vertices and values v are directed weights
+-- of the edge  p --> q where v is a list of numbers such that matrix{v} * basis(1,G.HTpt)
+-- gives the weight
+--G.HTpt : a polynomial ring, represent the T-equivariant cohomology ring of a point
+
+
+MomentGraph = new Type of MutableHashTable
```

Comment:
Since you have a cache table, maybe this doesn't have to be mutable.   Then you could compare two moment graphs with "===".

Reply:
MomentGraph is a HashTable now.